### PR TITLE
Debugger improvements

### PR DESCRIPTION
**Pull request by TGEnigma**
I should probably note that I don't usually program in C++, as I mainly use C# so I'd appreciate any advice or suggestions you can give me to improve my code.

Changes: (copied from commit description) 

* Added 'Step Over' functionality
 * Added special SPU pause functionality that pauses the SPU thread when the tag mask is at 0x80000000 by holding ctrl while pausing
  * Go to address dialog now evaluates expressions, including defined variables such as pc, r1, r2, etc
    - Requires QtScript to be linked with the project
  * Made the option to center shown addresses (Go to addr/pc) optional by making it an entry in the GUI ini config
  * Shown addresses now appear 'selected'
  * New keyboard shortcuts!
    - Ctrl+G -> Go to address
    - F10 -> Step Over
    - F11 -> Step (Into)

### CODE REVIEW

**Comment by VelocityRa** [comment]
Really nice.
Looks like most `rpcs3/rpcs3.vcxproj` changes are irrelevant, needs cleaning up.

**Comment by elad335** [comment]
squash the commits but other than that ,changes are awesome

**Comment by TGEnigma** [comment]
Yeah sorry about that, I had some trouble getting things to work and accidentally left in some things I changed while I was tinkering with it.

**Comment by elad335** [comment]
@TGEnigma do a `git rebase -i head~2` to merge the two commots into one and then push force

**Comment by flash-fire** [comment]
Travis is upset because you didn't link in qtscript in cmake.  There's other examples of how to do this that you should be able to find easy enough.  But, if you just want the end code, message me (aerodynamic chest) on discord ;)

**Comment by TGEnigma** [comment]
Can somebody help try to get Travis to build? I'm out of ideas.

**Comment by hcorion** [comment]
Strange issue with travis, would it be possible to just not use QtScript? It's a deprecated feature.

**Comment by TGEnigma** [comment]
It might be deprecated but it works fine for our intents and purposes. Cutting features instead of fixing the underlying problem seems like a bad idea to me IMO.

**Comment by danilaml** [comment]
@TGEnigma have you looked at http://doc.qt.io/qt-5/qjsengine.html ? This seems to be the proper (not deprecated) way to go forward,

**Comment by hcorion** [comment]
I agree with danilaml, it seems quite easy to do what you want with qjsengine.

**Comment by TGEnigma** [comment]
I'll look into it, I hadn't before.

**Comment by jarveson** [comment]
Good additions! Only thing is i'd be hesitant about the spu stall in this form, theres no rhyme or reason to the way tag masks are assigned or used in games, and its partial luck that 'that game' benefits from the pause of that mask. This will only cause faqs and tutorials about performance using this feature for possible extra performance. 

One other suggestion, if your up for more debugger improvements, in this pr or another, is the highlighting/coloring of the debug tab instructions, i think megamouse setup some system (or maybe it was all qt) to highlight cgdisasm which would be cool to get in the ppu/spu debugger

**Comment by TGEnigma** [comment]
If you think it'd be better if I took out the spu pause utility then I will, I just added it because it's a common use case right now.
I intend to start a larger overhaul of the debugger at some point to which I would like to add function coloring and integrate the memory viewer (for editing & bps) similarly to how it's implemented in other emulators.

**Comment by Asinin3** [comment]
There is already guides on pausing threads with p5 i don't really see an issue with making it a bit easier to do, though from my understanding if spu loop detection worked perfectly then there would be no real reason to do it. 

The syntax highlighting megamouse added to cgdisasm was done here: https://github.com/RPCS3/rpcs3/pull/3550

**Comment by AniLeo** [comment]
IMO it's also not ideal to just have it pause 0x8000000 just because P5 runs better with some threads stopped. At least make it a setting to allow one to set what tag mask to stop at.

**Comment by TGEnigma** [comment]
Are any other games even known to benefit from pausing SPU threads?

**Comment by AniLeo** [comment]
Probably not, a reason to leave that auto tag thing out I'd say

**Comment by TGEnigma** [comment]
From testing it seems to be a good successful way to pause SPU threads in games besides P5.
I'll leave it up to Neko to decide whether to keep it or not.

**Comment by hcorion** [comment]
Because I'd like to have the debugger improvements in master, would it be possible to just remove it, and do the controversial spu thread pausing in a different PR @TGEnigma ?

**Comment by TGEnigma** [comment]
@hcorion I removed it.

**Comment by scribam** [comment]
@TGEnigma Would you mind resolving conflicting files?

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 1]
```diff
@@ -397,72 +421,131 @@ void debugger_frame::Show_Val()
...
+void debugger_frame::DoStep(bool stepOver)
 {
 	if (const auto cpu = this->cpu.lock())
 	{
-		if (test(cpu_flag::dbg_pause, cpu->state.fetch_op([](bs_t<cpu_flag>& state)
-		{
-			state += cpu_flag::dbg_step;
-			state -= cpu_flag::dbg_pause;
-		})))
+		if (test(cpu_flag::dbg_pause, cpu->state))
```
Comment (Nekotekina) [inline rpcs3/rpcs3qt/debugger_frame.cpp]: You have replaced atomic op with non-atomic

**Comment by TGEnigma** [comment]
I fixed the issues you mentioned, but I accidentally rebased it on top of the original commit instead of making a new one.

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 2]
```diff
@@ -397,72 +420,127 @@ void debugger_frame::Show_Val()
...
+	scriptEngine.globalObject().setProperty("pc", GetPc());
+
+	if (thread->id_type() == 1)
+	{
+		auto ppu = static_cast<ppu_thread*>(thread.get());
+
+		for (int i = 0; i < 32; ++i)
+			scriptEngine.globalObject().setProperty(QString("r%1").arg(i), QJSValue((double)(ppu->gpr[i])));
+
+		scriptEngine.globalObject().setProperty("lr", QJSValue((double)ppu->lr));
```
Comment (Nekotekina) [inline rpcs3/rpcs3qt/debugger_frame.cpp]: Double type cannot encode full range of 64-bit unsigned value. Is it really what you meant?

Comment (TGEnigma) [reply rpcs3/rpcs3qt/debugger_frame.cpp]: There is no alternative

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 3]
```diff
@@ -531,22 +618,35 @@ void debugger_frame::OnBreakpointList_delete()
 	}
 }
 
-debugger_list::debugger_list(debugger_frame* parent) : QListWidget(parent)
+debugger_list::debugger_list(debugger_frame* parent, std::shared_ptr<gui_settings> settings) : QListWidget(parent)
```
Comment (danilaml) [inline rpcs3/rpcs3qt/debugger_frame.cpp]: Don't pass shared_ptr by value. Use const reference. It looks like this error is a common occurrence in this code (i.e. not only in this file). Would you mind fixing it? Also, I've noticed that it was added by @Megamouse in this commit: https://github.com/RPCS3/rpcs3/commit/ad66dbfd0b4db8cdb1c312e4bd8e08515c290663 (well, unless the settings can libe even after rpcs3_app is destroyed/closed). 

Doesn't seem like there is any need for shared_ptr at all (though in this case the overhead is negligible).

Comment (Megamouse) [reply rpcs3/rpcs3qt/debugger_frame.cpp]: @danilaml fyi: pirhoman is going to refactor this file. He has to wait for this Pull request to get merged. So he can also make this change later in favor for a faster merge now.

**Pull request merged**